src/util/address.rs

```diff
@@ -135,7 +135,100 @@ impl From<bech32::Error> for Error {
     }
 }
 
-/// The different types of addresses.
+/// Detailed address information returned by [`Address::address_info`]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+// #[non_exhaustive]
+pub enum AddressInfo {
+    /// Witness versions recognized by the current consensus
+    Known(AddressType),
+    /// Future witness versions (>1)
+    Future(WitnessVersion, /** witness program length */ usize),
+    /// Witness version is 0 and is known to consensus, but witness program has non-standard length
+    /// which is not supported by consensus rules and can't be spent.
+    NotSpendableWitnessV0(/** witness program length */ usize),
+    /// Witness version is 1 and is known to the consensus rules, but witness program has custom
+    /// length and reserved for the future use outside of Taproot
+    NonTaprootWitnessV1(/** witness program length */ usize),
+}
+
+impl From<AddressType> for AddressInfo {
+    #[inline]
+    fn from(address_type: AddressType) -> Self {
+        AddressInfo::Known(address_type)
+    }
+}
+
+
+impl AddressInfo {
+    /// Constructs pay-to-pubkey-hash variant
+    #[inline]
+    pub fn p2pkh() -> AddressInfo {
+        AddressInfo::Known(AddressType::P2pkh)
+    }
+
+    /// Constructs pay-to-script-hash variant
+    #[inline]
+    pub fn p2sh() -> AddressInfo {
+        AddressInfo::Known(AddressType::P2sh)
+    }
+
+    /// Constructs pay-to-witness-pubkey-hash variant
+    #[inline]
+    pub fn p2wpkh() -> AddressInfo {
+        AddressInfo::Known(AddressType::P2wpkh)
+    }
+
+    /// Constructs pay-to-witness-script-hash variant
+    #[inline]
+    pub fn p2wsh() -> AddressInfo {
+        AddressInfo::Known(AddressType::P2wsh)
+    }
+
+    /// Constructs pay-to-pubkey-hash variant
+    #[inline]
+    pub fn p2tr() -> AddressInfo {
+        AddressInfo::Known(AddressType::P2tr)
+    }
+
+    /// Returns known address type [`AddressType`], if any
+    #[inline]
+    pub fn address_type(self) -> Option<AddressType> {
+        match self {
+            AddressInfo::Known(address_type) => Some(address_type),
+            AddressInfo::Future(_, _) => None,
+            AddressInfo::NonTaprootWitnessV1(_) => None,
+            AddressInfo::NotSpendableWitnessV0(_) => None,
+        }
+    }
+
+    /// Detects if the address can be spend only with a valid signature (i.e. not anyone can spend
+    /// address of a future witness version or deterministically unspendable)
+    #[inline]
+    pub fn requires_valid_signature(self) -> bool {
```

Comment:
`requires_valid_signature` is not a good name because there may be p2sh/p2wsh addresses that don't require valid signatures to spend. How about calling this `is_known`? Note that is the same as `address_type().is_some()`. I wonder if it is worth having an API for it.

Reply:
This is also the same as the already existing `is_standard`.